external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -73,12 +73,19 @@ class DirectKafkaInputDStream[
 
   protected val kc = new KafkaCluster(kafkaParams)
 
-  protected val maxMessagesPerPartition: Option[Long] = {
-    val ratePerSec = context.sparkContext.getConf.getInt(
+  private val ratePerSec: Int = context.sparkContext.getConf.getInt(
       "spark.streaming.kafka.maxRatePerPartition", 0)
-    if (ratePerSec > 0) {
+  protected def maxMessagesPerPartition: Option[Long] = {
+    val estimatedRate = rateController.map(_.getLatestRate().toInt).getOrElse(-1)
+    val numPartitions = currentOffsets.keys.size
+    val effectiveRatePerSec = if (estimatedRate > 0) {
+      (ratePerSec min (estimatedRate / numPartitions))
```

Comment:
`ratePerSec min (estimatedRate / numPartitions)`.May be calling is postfix was wrong, but we generally do not use the dot-less notation unless the later term is a function body. So we do use 
```
option.getOrElse {
   anotherResult
}
```
But generally dont use 
`x max y`